### PR TITLE
[WebInterface][DataCollector] Add progression info

### PR DESCRIPTION
**Pull request by valouvaliavlo**

### CODE REVIEW

**Comment by valouvaliavlo** [comment]
https://feedback.octobot.online/posts/77/backtest-download-percentage
can be marked as done

File: Backtesting/collectors/exchanges/exchange_history_collector/history_collector.py [hunk 1]
```diff
@@ -73,16 +73,20 @@ async def start(self):
...
                 await self.get_ticker_history(self.exchange_name, symbol)
                 await self.get_order_book_history(self.exchange_name, symbol)
                 await self.get_recent_trades_history(self.exchange_name, symbol)
 
-                for index, time_frame in enumerate(self.time_frames):
+                for time_frame_index, time_frame in enumerate(self.time_frames):
+                    self.current_time_frame_index = (symbol_index * len(self.time_frames)) + time_frame_index + 1
                     self.logger.info(
-                        f"[{index}/{len(self.time_frames)}] Collecting {symbol} history on {time_frame}...")
+                        f"[{time_frame_index}/{len(self.time_frames)}] Collecting {symbol} history on {time_frame}...")
```
Comment (GuillaumeDSM) [inline Backtesting/collectors/exchanges/exchange_history_collector/history_collector.py:89]: 👍

File: Backtesting/collectors/exchanges/exchange_history_collector/history_collector.py [hunk 2]
```diff
@@ -152,8 +158,8 @@ async def get_ohlcv_history(self, exchange, symbol, time_frame):
             while since < last_candle_timestamp if not self.end_timestamp \
                     else (last_candle_timestamp < self.end_timestamp - (time_frame_sec * 1000)):
                 since = last_candle_timestamp
-                progress = (since-start_fetch_time) / total_interval
-                self.logger.info(f"[{round(progress *100)}%] historical data fetched for {symbol} {time_frame}")
+                self.current_time_frame_percent = round((since-start_fetch_time) / total_interval * 100)
```
Comment (GuillaumeDSM) [inline Backtesting/collectors/exchanges/exchange_history_collector/history_collector.py:161]: 👍

File: Services/Interfaces/web_interface/__init__.py [hunk 3]
```diff
@@ -126,6 +127,8 @@ def send_general_notifications(**kwargs):
 def send_backtesting_status(**kwargs):
     _send_notification(BACKTESTING_NOTIFICATION_KEY, **kwargs)
 
+def send_data_collector_status(**kwargs):
+    _send_notification(DATA_COLLECTOR_NOTIFICATION_KEY, **kwargs)
```
Comment (GuillaumeDSM) [inline Services/Interfaces/web_interface/__init__.py:131]: 👍

File: Services/Interfaces/web_interface/models/backtesting.py [hunk 4]
```diff
@@ -130,46 +130,69 @@ def get_delete_data_file(file_name):
...
+            current_time_frame_index, total_time_frame, current_time_frame_percent = \
+                            backtesting_api.get_data_collector_progress(data_collector)
+            progress["current_time_frame_index"] = current_time_frame_index
+            progress["total_time_frame"] = total_time_frame
+            progress["current_time_frame_percent"] = current_time_frame_percent
+            return "collecting", progress
+        if backtesting_api.is_data_collector_finished(data_collector):
+            return "finished", progress
+        return "starting", progress
+    return "not started", progress
```
Comment (GuillaumeDSM) [inline Services/Interfaces/web_interface/models/backtesting.py:147]: 👍

File: Services/Interfaces/web_interface/static/js/common/data_collector_util.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    else{
+        lock_collector_ui(false);
+    }
+    if(first_refresh_state === ""){
+        first_refresh_state = data_collector_status;
+    }
+}
+
+let first_refresh_state = "";
```
Comment (GuillaumeDSM) [inline Services/Interfaces/web_interface/static/js/common/data_collector_util.js:65]: what is `first_refresh_state ` used for ?

File: Services/Interfaces/web_interface/web.py [hunk 6]
```diff
@@ -41,7 +41,8 @@ class WebInterface(services_interfaces.AbstractWebInterface, threading.Thread):
     tools = {
         constants.BOT_TOOLS_BACKTESTING: None,
         constants.BOT_TOOLS_BACKTESTING_SOURCE: None,
-        constants.BOT_TOOLS_STRATEGY_OPTIMIZER: None
+        constants.BOT_TOOLS_STRATEGY_OPTIMIZER: None,
+        constants.BOT_TOOLS_DATA_COLLECTOR: None
```
Comment (GuillaumeDSM) [inline Services/Interfaces/web_interface/web.py:45]: 👍

File: Services/Interfaces/web_interface/websockets/data_collector.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    @websockets.websocket_with_login_required_when_activated
+    def on_connect(self):
+        super().on_connect()
+        self.on_data_collector_status()
+
+
+notifier = DataCollectorNamespace('/data_collector')
+web_interface.register_notifier(web_interface.DATA_COLLECTOR_NOTIFICATION_KEY, notifier)
+websockets.namespaces.append(notifier)
```
Comment (GuillaumeDSM) [inline Services/Interfaces/web_interface/websockets/data_collector.py:52]: 💯

**Review by GuillaumeDSM — COMMENTED** [review]
great job ! 💯 🏆

**Review by Herklos — COMMENTED** [review]
Very good job @valouvaliavlo!

**Review by GuillaumeDSM — APPROVED** [review]

**Review by Herklos — APPROVED** [review]

**Comment by GuillaumeDSM** [comment]
requires https://github.com/Drakkar-Software/OctoBot-Backtesting/pull/140

File: Backtesting/collectors/exchanges/exchange_history_collector/history_collector.py [hunk 8]
```diff
@@ -182,4 +188,5 @@ async def get_kline_history(self, exchange, symbol, time_frame):
         pass
 
     async def get_first_candle_timestamp(self, symbol, time_frame):
-        return (await self.exchange.get_symbol_prices(symbol, time_frame, limit=1, since=0))[0][commons_enums.PriceIndexes.IND_PRICE_TIME.value]
+        return (await self.exchange.get_symbol_prices(symbol, time_frame, limit=1, since=0))[0]
+                                            [commons_enums.PriceIndexes.IND_PRICE_TIME.value]
```
Comment (GuillaumeDSM) [inline Backtesting/collectors/exchanges/exchange_history_collector/history_collector.py:192]: invalid code :)

Comment (valouvaliavlo) [reply Backtesting/collectors/exchanges/exchange_history_collector/history_collector.py:192]: it's solved ! thanks

**Pull request merged**